rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+
+At the most basic level, the `Form` primitive is a regular Radix primitives package which expose a component split in multiple parts like other primitives. It offers a declarative way to use controls and labels, as well as a validation rules that are built on top of the native browser [constraint validation API](https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation).
+
+It abstracts away all of the complexity of accessibility and validation, using standard HTML attributes, as well as aria attributes.
+
+### API example
+
+Let's look at an example:
+
+```jsx
+import * as React from 'react';
+import * as Form from '@radix-ui/react-form';
+
+function Page() {
+  return (
+    <Form.Root
+      onSubmit={(event) => {
+        // `onSubmit` only triggered if it passes client-side validation
+        const data = Object.fromEntries(new FormData(event.currentTarget));
+        console.log(data);
+        event.preventDefault();
+      }}
+    >
+      <Form.Field name="name">
+        <Form.Label>Full name</Form.Label>
+        <Form.Control /> {/* renders a `input[type="text"]` by default */}
+        <Form.ClientMessage type="missingValue">Please enter your name.</Form.ClientMessage>
+      </Form.Field>
+
+      <Form.Field name="email">
+        <Form.Label>Email address</Form.Label>
+        <Form.Control type="email" />
+        <Form.ClientMessage type="missingValue">Please enter your email.</Form.ClientMessage>
+        <Form.ClientMessage type="typeMismatch">Please provide a valid email.</Form.ClientMessage>
+      </Form.Field>
+
+      <Form.Field name="age">
+        <Form.Label>Age</Form.Label>
+        <Form.Control type="number" min="18" max="57" step="1" />
+        <Form.ClientMessage type="missingValue">Please enter your age.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeOverflow">You must be 18 or older.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeUnderflow">You must be 57 or younger.</Form.ClientMessage>
+        <Form.ClientMessage type="stepMismatch">Please enter a whole number.</Form.ClientMessage>
+      </Form.Field>
+
+      <Form.Submit>Submit</Form.Submit>
+    </Form.Root>
+  );
+}
+```
+
+Note that there is not state present about the client, everything is uncontrolled by default (yet can be controlled). Also note that there are no attributes necessary to handle accessibility, this is all handled by the primitive:
+
+- label and controls are associated using the `name` provided on `Form.Field`.
+- when one or more client-side error message display, they are automatically associated to their matching control and focus is moved to the first invalid control.
+
+### Styling
+
+Similar to other Radix primtives, the `Form` primitive is unstyled by default. This means that you can style it using any CSS solution of your choice. Each part is a node you can style. Like other primitives, data attributes are present to aid styling states such as `data-invalid` and `data-valid`.
```

Comment:
Would it also apply an `aria-invalid` styling hook? I know Radix components so far seem to exclusively expose data attributes for state styling, any particular reason for that if matching aria-states already exists?

Reply:
There should be no need for `aria-invalid` as `Form.Control` renders an `input` so the native validity already works. If you are composing a custom control, we would expect the responsibility of making sure that component is accessible belongs in its implementation (similar to how we do with some of our primitives). This way, as long as the custom control relies on some native control internally, it should all work fine.

Reply:
Styling using `:invalid` is problematic since it will apply immediately when a form control is rendered, even before the user has interacted with it. An empty input with `required` will be invalid from the time it's rendered.

`:user-invalid` is designed to fix this and only apply after a user has interacted with the form, and has been selected as a priority for [Browser Interop 2023](https://github.com/web-platform-tests/interop/issues/178) but is still not supported in Chrome. `aria-invalid` is [specced](https://github.com/whatwg/html/issues/7848) slightly differently to `:invalid`/`element.validity.valid`, and more closely matches how `:user-invalid` works.

This is something you'll have to decide for your validation logic as well - should a message be shown immediately when the form is invalid (a required input will report `input.validity.valid === false` directly when it's rendered), or should it only happen after "significant user interaction" as with the `:user-invalid` selector.

Reply:
Yes, that's why we expose data attributes, as it's less reliant on implementation details (and also for this case you are mentioning in particular relating to `:invalid`).